### ISSUE TITLE
mpOpenAPI-2.0 limitation: only one web module is processed

### ISSUE THREAD

**Issue by Azquelt** [OpenLiberty/open-liberty#15032]
Currently, mpOpenAPI-2.0 only processes one web module in the first application to be deployed. This appears to be a limitation carried over from mpOpenAPI-1.1.

This limitation needs to be removed or documented.

Here are some examples of the behaviour:
 - one .war is deployed - an OpenAPI document is generated for that .war
 - two .wars are deployed - an OpenAPI document is generated for one of the .wars
 - an .ear containing two .wars is deployed - an OpenAPI document is generated for one of the .wars in the .ear

**Comment by tevans78** [comment]
Arthur and Artur confirmed that only openapi-3.1 supports multiple apps. mpOpenAPI only supports a simple app. Multiple modules are not currently supported by either.

**Comment by Thihup** [comment]
Is there a plan to fix it or at least allow to specify which application the OpenAPI should use?
Is very unpleasant to have a 'random' WAR generating the API if we are using an EAR or multiple wars.

**Comment by tevans78** [comment]
Some high-level options:
1. Work with the MicroProfile community to clarify the spec prior to potentially fixing this in the next version
2. Fix in Open Liberty to support multiple single module apps
3. Fix in Open Liberty to support multiple apps and multiple modules
4. As Thihup suggests, allow the single application/module to be specified
5. Document the limitation

**Comment by Azquelt** [comment]
We should be able to generate a document for each web module and then merge them.

There are a few places where we may need to make modifications to the documents when they are merged

- paths need to have the context root of the module prepended
  - currently paths are generated relative to the context root, and then the URL we generate for the `servers` element includes the context root
- paths from one module which are masked by the context root of another module should be removed to avoid the possibility of clashes between paths when the documents are merged
- items under `components` need to be renamed if they clash
  - all references to them need to be updated to match
- `operationId`s need to to be renamed if they clash as it's a requirement that it is unique within the document
  - all references to them need to be updated to match
  - this is not ideal as the spec states that this value may be used by tool and libraries to uniquely identify the operation
- the `info` object must be replaced (perhaps unless the info objects are identical across the merged applications)
  - this object includes the title, description, license and contact details for the API described, so it would be incorrect to merge it across multiple applications
- items under `securitySchemes` need to be renamed if they clash
  - all references need to be updated to match
- `externalDocs` should be removed
  - it only allows linking to one URL which will no longer be correct if we have merged documents from multiple applications
- Not sure how best to handle tags
  - Tags appear to only be used to group operations in generated documentation, so they don't seem critical
  - We could either rename any that clash, or we could assume that tags are likely to mean the same thing, even across multiple applications and so merging them makes more sense.

In addition, given the caveats of merging listed above, it might make sense to also provide documents for individual applications at separate URLs (e.g. `/openapi/<appname>` and `/openapi/<appname>/ui`) so that users can still retrieve the document for a single application if they need it.

**Comment by Azquelt** [comment]
This is available in the 21.0.0.12-beta available from https://openliberty.io/start/#runtime_betas

or from maven central (e.g. for use with the [maven][maven-plugin] or [gradle][gradle-plugin] plugins:
```
  <groupId>io.openliberty.beta</groupId>
  <artifactId>openliberty-runtime</artifactId>
  <version>21.0.0.12-beta</version>
  <type>zip</type>
```

You need to configure it to merge documentation for all applications by setting the MP Config property `mp.openapi.extensions.liberty.merged.include=all`, e.g. by adding that to your `bootstrap.properties`.

[maven-plugin]: https://github.com/OpenLiberty/ci.maven/blob/main/docs/installation-configuration.md#using-maven-artifact
[gradle-plugin]: https://github.com/OpenLiberty/ci.gradle/blob/main/docs/installLiberty.md#dependencies

**Comment by Azquelt** [comment]
This is included in 22.0.0.1.
